박정근/Programmers/Week26/순위_검색.java

```diff
@@ -0,0 +1,120 @@
+                    case "chicken":
+                        idx[c] = 0;
+                        break;
+                    case "java":
+                    case "frontend":
+                    case "senior":
+                    case "pizza":
+                        idx[c] = 1;
+                        break;
+                    case "python":
+                        idx[c] = 2;
+                        break;
+                }
+            }
+            
+            info_sum[idx[0]][idx[1]][idx[2]][idx[3]][Integer.parseInt(st.nextToken())] += 1;
+        }
+        
+        // 누적 합 만들기 (특정 항목에서 X점 이상인 사람의 수)
+        for (int i = 0; i < 3; i++) {
+            for (int j = 0; j < 2; j++) {
+                for (int k = 0; k < 2; k++) {
+                    for (int l = 0; l < 2; l++) {
+                        for (int sc = 99999; sc >= 0; sc--) {
+                            info_sum[i][j][k][l][sc] += info_sum[i][j][k][l][sc+1];
+                        }
+                    }
+                }
+            }
+        }
+        
+        int q_idx = 0;
+        for (String q : query) {
+            StringTokenizer st = new StringTokenizer(q);
+            
+            int[] idx_start = new int[4];
+            int[] idx_end = {3, 2, 2, 2};
+            for (int c = 0; c < 4; c++) { 
+                String now;
+                if (c != 0) now = st.nextToken(); // 중간에 있는 and 빼기
+                now = st.nextToken();
+                
+                switch(now) {
+                    case "cpp":
+                    case "backend":
+                    case "junior":
+                    case "chicken":
+                        idx_start[c] = 0;
+                        idx_end[c] = 1;
+                        break;
+                    case "java":
+                    case "frontend":
+                    case "senior":
+                    case "pizza":
+                        idx_start[c] = 1;
+                        idx_end[c] = 2;
+                        break;
+                    case "python":
+                        idx_start[c] = 2;
+                        idx_end[c] = 3;
+                        break;
+                    case "-":
+                        idx_start[c] = 0;
+                        break;
+                }
+            }
+        
+            int sc = Integer.parseInt(st.nextToken()); // 점수 조건
+            int res = 0;
+            
+            // -1 이면 전체를 탐색할 수 있도록 했습니다. 
+            // 복잡해 보이지만, 한 쿼리마다 최대 3 x 2 x 2 x 2번의 순회로 답을 얻을 수 있습니다.
+            for (int i = idx_start[0]; i < idx_end[0]; i++) {
+                for (int j = idx_start[1]; j < idx_end[1]; j++) {
+                    for (int k = idx_start[2]; k < idx_end[2]; k++) {
+                        for (int l = idx_start[3]; l < idx_end[3]; l++) {
+                            res += info_sum[i][j][k][l][sc];
+                        }
+                    }      
+                }
+            }
```

Comment:
```suggestion
            for (int language = idx_start[0]; language < idx_end[0]; language++) {
                for (int position = idx_start[1]; position < idx_end[1]; position++) {
                    for (int level = idx_start[2]; level < idx_end[2]; level++) {
                        for (int food = idx_start[3]; food < idx_end[3]; food++) {
                            res += info_sum[language][position][level][position][sc];
                        }
                    }      
                }
            }
```

3 차원 이상의 배열을 사용한 경험이 없어서 이렇게 많이 중첩되는 반복문을 본 적은 없는데, 각 인덱스 별로 의미를 나뉘어질 수 있도록 변수명을 설정하면 어떨까 생각 해봤습니다.

Reply:
+ 누적합 배열을 만든 것 까지 이해 했는데, 답이 어떻게 도출 되는건지 이해가 잘 안되서(제가) 조금 더 추가 설명 부탁 드려요!!

```java
prefix  100 | 100 | 90 | 60 | 10 | 0
score   10  |  20  | 30 | 40 | 50 | 60
```

이렇게 누적합이 있다면 타겟 점수가 30 이상 이라고 한다면 바로 info_sum[30] 에 해당하는 누적합을 확인 하면 된다고 생각 했는데 반복문이 돌아가니까 제가 잘 못 이해한 부분이 있는거 같은데 어딘지 못 찾겠네요.

Reply:
복잡한 중첩 반복문에서의 키워드 사용도 코드를 정리하는 데에 도움이 되는 것 같습니다! 앞으로는 이렇게 구현해야겠어요 😀

Reply:
prefix 가 어떤 걸 말하는지 잘 모르겠네요. 먼저 info_sum 의 의도를 설명하면 도움이 될 수 있을 것 같습니다.

```
cpp frontend junior pizza 100점 받은 지원자 : info_sum[0][1][0][1][100] += 1;
java backend senior chicken 50점 받은 지원자 : info_sum[1][0][1][0][50] += 1;
cpp backend junior chicken 1000점 받은 지원자 : info_sum[0][0][0][0][100] += 1;
```

info_sum의 인덱스는 각각 개발언어, 직군, 경력, 소울푸드, 코딩테스트 점수를 각각 인덱스화 시킨 것 입니다.
누적 합을 만들기 전에, info 배열에서의 데이터를 5차원 배열에 쌓으면서, 특정 항목에 해당하는 사람의 수가 몇 명인지를 info_sum 에 먼저 쌓았다고 보시면 됩니다. 만약 쿼리에서 특정 개발언어, 직군, 경력, 소울푸트, 코딩테스트 점수를 받은 사람의 수를 구한다고 하면 `info_sum[lan][pos][lv][food][score]` 를 출력하면 되게끔이요.

순위 검색 문제의 쿼리에서는 X점 이상 받은 사람을 구하기 때문에, 저는 **항목은 그대로 두고 점수만 누적 합 배열로 만들어서**, `info_sum[lan][pos][lv][food][score]` 를 접근하면 lan, pos, lv, food 를 선택한 사람 중 **score점 이상 받은 사람**을 구할 수 있게끔 했습니다. 사실상 4차원 배열까지는 그냥 항목에 접근한거고, 누적 합으로 만들어진건 마지막 5차원 배열입니다. 다음 코드에서 볼 수 있습니다.

```
// 누적 합 만들기 (특정 항목에서 X점 이상인 사람의 수)
for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 2; j++) {
        for (int k = 0; k < 2; k++) {
            for (int l = 0; l < 2; l++) {
                for (int sc = 99999; sc >= 0; sc--) { 
                    // 99999점 이상인 사람 수는 99999 + 100000
                    // 99998점 이상인 사람 수는 99998 + (99999 + 100000)
                    // ...
                    info_sum[i][j][k][l][sc] += info_sum[i][j][k][l][sc+1];
                }
            }
        }
    }
}
```

특정 항목에만 쿼리가 주어진다면 여기서 끝이지만, 쿼리 중에서는 특정 항목의 전체를 보아야 하는 `-` 도 존재하기 때문에, 쿼리를 분류할 때 각 항목마다 탐색 시작 인덱스와 종료 인덱스를 설정했습니다.

```
int[] idx_start = new int[4];
int[] idx_end = {3, 2, 2, 2};
for (int c = 0; c < 4; c++) { 
    String now;
    if (c != 0) now = st.nextToken(); // 중간에 있는 and 빼기
    now = st.nextToken();
    
    switch(now) {
        case "cpp":
        case "backend":
        case "junior":
        case "chicken":
            idx_start[c] = 0;
            idx_end[c] = 1;
            break;
        case "java":
        case "frontend":
        case "senior":
        case "pizza":
            idx_start[c] = 1;
            idx_end[c] = 2;
            break;
        case "python":
            idx_start[c] = 2;
            idx_end[c] = 3;
            break;
        case "-":
            idx_start[c] = 0;
            break;
    }
}
```

idx_end 배열은 초기화할 때 5차원 배열의 맨 끝 인덱스로 잡아둔 뒤, `-`가 들어온다면 시작 인덱스를 0으로 설정해 해당 항목을 처음부터 끝까지 탐색할 수 있도록 했습니다. 그 외 특정 항목이 정해져서 들어온다면 idx_start 와 idx_end 가 한 항목만 탐색할 수 있도록 설정됩니다.

이렇게 설정하면, 쿼리에 `-` 가 들어왔을 때 반복문에서는 해당 항목의 전체를 탐색해 더합니다. 경력이 `-` 라면 `junior`와 `senior` 를 전부 더하고, 직군과 경력이 `-` 라면 `frontend, junior` `frontend, senior` `backend, junior` `backend, senior` 를 전부 더합니다.
```
java and backend and - and pizza 100 : 
(java, backend, junior, pizza, 100<=), (java. backend, senior, pizza, 100<=) 탐색

java and - and - and pizza 100 : 
(java, frontend, junior, pizza, 100<=), (java. frontend, senior, pizza, 100<=)
(java, backend, junior, pizza, 100<=), (java. backend, senior, pizza, 100<=) 탐색
```

해당 원리로 쿼리가 수행됩니다.

Reply:
prefix 는 prefix sum 이라고 말했어야 했는데 제가 잘 못 말했네요.
설명 감사합니다 ~

제가 누적합 개념을 이해 못했던 이유는 인덱스 0 부터 쌓는 누적합이랑 헷갈려서 이해를 못 했던것 같습니다.
지금 보니까 뒤에서 부터 누적합을 쌓아서 `특정 점수 이상` 사람들에 대한 정답을 도출할 수 있도록 되어있네요.